core/lib/zksync_core/src/eth_watch/event_processors/set_chain_id.rs

```diff
@@ -0,0 +1,98 @@
+use zksync_dal::StorageProcessor;
+use zksync_types::{web3::types::Log, Address, ProtocolUpgrade, H256};
+
+use crate::eth_watch::{
+    client::{Error, EthClient},
+    event_processors::EventProcessor,
+    metrics::{PollStage, METRICS},
+};
+
+/// Responsible for saving new protocol upgrade proposals to the database.
+#[derive(Debug)]
+pub struct SetChainIDEventProcessor {
+    diamond_proxy_address: Address,
+    set_chain_id_signature: H256,
+}
+
+impl SetChainIDEventProcessor {
+    pub fn new(diamond_proxy_address: Address) -> Self {
+        Self {
+            diamond_proxy_address,
+            set_chain_id_signature: H256::default(),
+            // TODO: uncomment when we merge the Shared Bridge contracts.
+            // state_transition_manager_contract()
+            //     .event("SetChainIdUpgrade")
+            //     .expect("SetChainIdUpgrade event is missing in ABI")
+            //     .signature(),
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl EventProcessor for SetChainIDEventProcessor {
+    async fn process_events(
+        &mut self,
+        storage: &mut StorageProcessor<'_>,
+        client: &dyn EthClient,
+        events: Vec<Log>,
+    ) -> Result<(), Error> {
+        let mut upgrades = Vec::new();
+        let events_iter = events.into_iter();
+
+        let filtered_events = events_iter.filter(|log| {
+            log.topics[0] == self.set_chain_id_signature
+                && log.topics[1] == self.diamond_proxy_address.into()
+        });
+
+        // SetChainId does not go through the governance contract, so we need to parse it separately.
+        for event in filtered_events {
+            let timestamp = client
+                .get_block(event.block_hash.expect("event without block_hash"))
+                .await?
+                .expect("event's block not found")
+                .timestamp;
+            let upgrade = ProtocolUpgrade::decode_set_chain_id_event(event, timestamp.as_u64())
```

Comment:
It seems you don't need whole `ProtocolUpgrade`. I think `decode_set_chain_id_event` should return only `ProtocolVersionId` and upgrade tx. Then won't need nor querying timestamp, nor `get_block` method in client interface